[VPlan] Expand VPWidenIntOrFpInductionRecipe into separate recipes

lukel97 opened the pull request:
The motivation of this PR is to make #115274 easier to implement, and should allow us to add EVL support by just passing EVL to the VF operand.

The current difficulty with widening IVs with EVL is that VPWidenIntOrFpInductionRecipe generates its own backedge value. Since it's a VPHeaderPHIRecipe the VF operand must be in the preheader, which means we can't use the EVL since it's defined in the loop body.

The gist in this PR is to take the approach in #114305 and expand VPWidenIntOrFpInductionRecipe into several recipes for the initial value, phi and backedge value just before execution. I.e. this example:

```
  vector.ph:
  Successor(s): vector loop

  <x1> vector loop: {
    vector.body:
      WIDEN-INDUCTION %i = phi %start, %step, %vf
      ...
      EMIT branch-on-count ...
    No successors
  }
```

gets expanded to:

``` 
vector.ph:
  ...
  vp<%induction.start> = ...
  vp<%induction.increment> = ...

Successor(s): vector loop

<x1> vector loop: {
  vector.body:
    ir<%i> = WIDEN-PHI vp<%induction.start>, vp<%vec.ind.next>
    ...
    vp<%vec.ind.next> = add ir<%i>, vp<%induction.increment>
    EMIT branch-on-count ...
  No successors
}
```

This allows us to a value defined in the loop in the backedge value, and also means we can just reuse the existing backedge fixups in VPlan::execute without having to specially handle it ourselves.

After this #115274 should just become a matter of setting the VF operand to EVL (and building the increment step in the loop body, not the preheader).

lukel97 committed:
[LV] Expand VPWidenIntOrFpInductionRecipe into separate recipes

lukel97 committed:
Remove VPWidenIntOrFpInductionPHIRecipe, use VPWidenPHIRecipe instead

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1043,29 +1043,19 @@ void VPlan::execute(VPTransformState *State) {
-        auto *GEP = cast<GetElementPtrInst>(State->get(WidenPhi));
-        Phi = cast<PHINode>(GEP->getPointerOperand());
-      }
+    if (isa<VPWidenPointerInductionRecipe>(&R)) {
+      auto *WidenPhi = cast<VPWidenPointerInductionRecipe>(&R);
```
```suggestion
    if (auto *WidenPhi = dyn_cast<VPWidenPointerInductionRecipe>(&R)) {
```

lukel97 replied:
Apologies, I didn't see this earlier. Fixed now

fhahn commented:
Overall direction aligns well with current and future VPlan roadmap. As is the current version adds a bunch of complexity though, will need some time to see how/if this could be further simplified

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1706,29 +1732,27 @@ void VPWidenIntOrFpInductionRecipe::execute(VPTransformState &State) {
-  // Add induction update using an incorrect block temporarily. The phi node
-  // will be fixed after VPlan execution. Note that at this point the latch
-  // block cannot be used, as it does not exist yet.
-  // TODO: Model increment value in VPlan, by turning the recipe into a
-  // multi-def and a subclass of VPHeaderPHIRecipe.
-  VecInd->addIncoming(LastInduction, VectorPH);
```
Could we first have a separate patch focused specifically on completing this TODO? If I understand correctly, this seems to be the primary issue at hand.

Additionally, I’d like to understand the difference between introducing a new recipe, `VPWidenIntOrFpInductionBackedgeRecipe`, and combining `VPInstruction::Mul` and `VPInstruction::Add`(https://github.com/llvm/llvm-project/pull/82021). What are the key distinctions, advantages, or trade-offs between these two approaches?

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1706,29 +1732,27 @@ void VPWidenIntOrFpInductionRecipe::execute(VPTransformState &State) {
-  // Add induction update using an incorrect block temporarily. The phi node
-  // will be fixed after VPlan execution. Note that at this point the latch
-  // block cannot be used, as it does not exist yet.
-  // TODO: Model increment value in VPlan, by turning the recipe into a
-  // multi-def and a subclass of VPHeaderPHIRecipe.
```
Could we first have a separate patch focused specifically on completing this TODO? If I understand correctly, this seems to be the primary issue at hand.

Additionally, I’d like to understand the difference between introducing a new recipe, `VPWidenIntOrFpInductionBackedgeRecipe`, and combining `VPInstruction::Mul` and `VPInstruction::Add`(https://github.com/llvm/llvm-project/pull/82021). What are the key distinctions, advantages, or trade-offs between these two approaches?

lukel97 replied:
> Could we first have a separate patch focused specifically on completing this TODO? If I understand correctly, this seems to be the primary issue at hand.

The VPHeaderPHIRecipe part of this todo is already done, but I'm not sure if the part about turning this into a multi-def will fix the issue. 

If I'm understanding this correctly, if we had a multi-def recipe we would still be executing the increment/backedge value in the header before the rest of the loop body, and so we wouldn't be able to use the EVL in it. 

> Additionally, I’d like to understand the difference between introducing a new recipe, VPWidenIntOrFpInductionBackedgeRecipe, and combining VPInstruction::Mul and VPInstruction::Add(https://github.com/llvm/llvm-project/pull/82021).

I actually initially tried to avoid the new recipes and just use VPInstructions but I struggled with trying to splat two scalar VPInstructions into a vector operand. But actually now that I realise that #82021 does it I'll give it another try!

I would prefer to avoid adding extra recipes if possible, I don't see any particular advantage to having explicit recipes.

fhahn replied:
Yes, in general at this late stage there's less advantage from specialized recipes, better to use simpler recipes

fhahn replied:
Yes, in general at this late stage there's less advantage from specialized recipes, better to use simpler recipes

arcbbb commented:
Appreciate the effort! Just wondering - for EVL transform, do you expect to generate vp intrinsics during the backedge recipe execution, or will a separate transformation be needed to convert the backedge recipes into WidenIntrinsicRecipes?

lukel97 commented:
> Appreciate the effort! Just wondering - for EVL transform, do you expect to generate vp intrinsics during the backedge recipe execution, or will a separate transformation be needed to convert the backedge recipes into WidenIntrinsicRecipes?

I'm trying to get rid of the backedge recipe and just replace it with regular VPInstructions + VPWidenRecipes. In that case we could then rerun transformRecipestoEVLRecipes after the expansion?

Although I just want to double check, I was thinking that doing the regular widening for now would still be correct. Can we convert it to VP intrinsics as a separate task after the EVL transform or is it needed for correctness?

As a side note on RISC-V, I wonder if RISCVLOptimizer.cpp from #108640 might take care of removing the VL toggle for us in the backend as well

npanchen commented:
> Although I just want to double check, I was thinking that doing the regular widening for now would still be correct. Can we convert it to VP intrinsics as a separate task after the EVL transform or is it needed for correctness?

I don't think regular widening of these inductions (i.e. without EVL) will be correct. The main problem is vsetvl's  6.3.2 so that if increment is VLMAX, last two iteration may result to overincrement.
fyi: QEMU has special [`rvv_vl_half_avl`](https://lists.nongnu.org/archive/html/qemu-devel/2024-08/msg01060.html) option to properly test 6.3.2

lukel97 commented:
> > Although I just want to double check, I was thinking that doing the regular widening for now would still be correct. Can we convert it to VP intrinsics as a separate task after the EVL transform or is it needed for correctness?
> 
> I don't think regular widening of these inductions (i.e. without EVL) will be correct. The main problem is vsetvl's 6.3.2 so that if increment is VLMAX, last two iteration may result to overincrement. fyi: QEMU has special [`rvv_vl_half_avl`](https://lists.nongnu.org/archive/html/qemu-devel/2024-08/msg01060.html) option to properly test 6.3.2

This will allow us to use EVL as the increment value in the backedge value to deal with vsetvl's 6.3.2, instead of VF, e.g. something like

```llvm
%phi.next = add <vscale x 2 x i32> %phi.cur, %evl
```

My thinking is that it shouldn't matter if it's a vp.add or a regular widened add since the phi value should be correct across all lanes for the next iteration?

lukel97 commented:
> However, it does make more sense to stick with same approach: either emit vp-intrinsics always out of vectorizer or to have a minimal subset + postprocessing pass after vectorizer. Current vision is to do former. By the way, unary and binary evl-recipes are already supported #93854, so it should be fine to use that, right ?

Agreed, that seems like a sensible way forwards. And just to echo what @Mel-Chen and @arcbbb are saying, we can do the EVL-recipe conversion as a separate incremental follow up to #115274 (which hopefully in turn can become a follow up to this PR)

lukel97 committed:
Remove initial and backedge recipes, expand out to regular VPInstructions + splat and step vector recipes

lukel97 commented:
I've reworked this to remove the IV specific initial value and backedge recipes, and instead they're modelled in VPlan during expansion.

I had to add two new recipes to model splats and step vectors in VPlan. I needed the former because although it looks like you can broadcast underlying scalar IR values, I couldn't see a way to broadcast a scalar VPValue with no underlying instruction.

We could also get rid of the PHI recipe if we relax the native-path restriction on VPWidenPHIRecipe, see #118662 

There's some extra diff now due to how this doesn't emit an identity add when building the initial value, I've also opened a separate PR to split this off: #119668

lukel97 commented:
> For the induction step, there's already code the materializes it (for unrolling), #119284 generalizes it to a VPInstruction opcode. Would that help?

I think I can reuse that here, that would take care of the casting part. And I guess we would get the benefit of the constant step optimization too. I'll rebase this PR on top of it

I think we would need to expand the VPWidenIntOrFpInductionRecipes first in convertToConcreteRecipes so that the WideIVSteps get a chance to be expanded afterwards?

lukel97 committed:
Only truncate the IV step, don't widen it

lukel97 committed:
Use recipe debugloc after 734a204fbd4b790048c57f79351ad8beeb1000ce

lukel97 committed:
Mark splat and step vector recipes as free from side effects/reads/writes

lukel97 committed:
Fix crash when pointer IV comes after intorfp IV

lukel97 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2622,6 +2622,19 @@ static void replaceVPBBWithIRVPBB(VPBasicBlock *VPBB, BasicBlock *IRBB) {
 
   VPBlockUtils::reassociateBlocks(VPBB, IRVPBB);
+
+  // Replace any references of VPBB in widened phi incoming blocks with IRBB.
+  for (auto *BB : IRVPBB->successors()) {
+    while (auto *Region = dyn_cast<VPRegionBlock>(BB))
+      BB = Region->getEntry();
+
+    for (auto &R : *cast<VPBasicBlock>(BB))
+      if (auto *WidenPhiR = dyn_cast<VPWidenPHIRecipe>(&R))
+        for (unsigned I = 0; I < WidenPhiR->getNumOperands(); I++)
+          if (WidenPhiR->getIncomingBlock(I) == VPBB)
+            WidenPhiR->setIncomingBlock(I, IRVPBB);
+  }
+
```
After c7ebe4fd0afadcddd53ec89e6030f9a8f5370e1f we will crash in fixNonInductionPHIs if we use a VPWidenPHIRecipe with the vector preheader as an incoming block, because the phi will reference the old non-IRBB vector preheader.

The existing native vplan tests don't seem to hit this because they don't create any widened phis that have the vector preheader/scalar preheader/middle block as an incoming block.

arcbbb replied:
Great find! Tracking this type of use is difficult as VPBasicBlock isn’t a VPValue. It seems we should handle all such uses within reassociateBlocks.

lukel97 replied:
This should be split out now after #125481 and #126388

lukel97 committed:
Use dyn_cast

lukel97 commented:
I've updated this PR to remove the need for a custom phi recipe, and instead reuse VPWidenPHIRecipe, which also requires #120338.

Now there should only be two extra recipes needed here, one to model step vectors and another to model splats.

lukel97 committed:
Remove VPSplatRecipe, replace with VPInstruction::Splat

lukel97 committed:
Update comment

lukel97 committed:
Pass Name when generating VPInstruction::Splat

LiqinWeng commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -261,6 +261,14 @@ class VPBuilder {
+    return tryInsertInstruction(new VPInstruction(VPInstruction::Splat, {Val}));
+  }
+
+  VPStepVectorRecipe *createStepVector(Type *Ty) {
```
Is it possible to use VPInstruction::Step instead of VPStepVectorRecipe?

lukel97 replied:
I tried to create a `VPInstruction::StepVector` but unfortunately it looks like we need a recipe to encode the scalar type, since there's no operands. Unless there's a way to plumb it through the VPInstruction? I think https://github.com/llvm/llvm-project/pull/119284 might have done something similar here by passing through a dummy constant

fhahn replied:
Yep I will follow-up to the comments in https://github.com/llvm/llvm-project/pull/119284 in the next few days.

lukel97 replied:
I've reworked this to be a VPInstruction::StepVector now

lukel97 committed:
Allow splatting VPScalarCastRecipe directly, remove VPInstruction::splat

lukel97 committed:
Remove VPStepVectorRecipe, replace with VPInstruction

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1887,10 +1887,148 @@ void VPlanTransforms::createInterleaveGroups(
 }
 
-void VPlanTransforms::convertToConcreteRecipes(VPlan &Plan) {
+/// Expand a VPWidenIntOrFpInduction into executable recipes. for the initial
```
Sure, just to check do you mean by checking against the dumped vplan in the debug output?

LiqinWeng replied:
yes~

lukel97 replied:
I've added a test in 856ac750f72eb3aff6deada6e01a7c92f15fadf3

lukel97 commented:
I've rebased and updated this to remove as many custom recipes and instructions as possible:

- The splat VPInstruction was removed and instead I reused the built in splatting mechanism in `VPTransformState::get`. I had to update it to handle splatting values that weren't instructions, and relax a `vputils::onlyFirstLaneUsed` assertion on VPScalarCastRecipe
- The VPStepVectorRecipe was removed and replaced with a VPInstruction that uses the same constant value trick in https://github.com/llvm/llvm-project/pull/119284 to specify the scalar type.

After https://github.com/llvm/llvm-project/pull/119284 lands I can rebase this on top of it to reuse it too.

lukel97 committed:
Add VPlan debug output test

lukel97 committed:
Undo stray whitespace change

lukel97 committed:
Fix ARM test

lukel97 commented:
Ping

lukel97 commented:
Gentle ping, all stacked patches should be landed now and rebased on main

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2013,10 +2013,147 @@ void VPlanTransforms::createInterleaveGroups(
+  if (WidenIVR->getNumOperands() == 5) {
+    Inc = WidenIVR->getSplatVFValue();
+    Prev = WidenIVR->getLastUnrolledPartOperand();
+    assert(Inc && Prev);
```
Add the assertion message

lukel97 replied:
I looked into the assert again and it didn't seem that necessary, so I just removed it in dff84ec397549bee30382847dc70d3f0513dee9b

lukel97 committed:
Remove unnecessary assert

lukel97 committed:
Merge remote-tracking branch 'origin/main' into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -265,6 +265,12 @@ class VPBuilder {
         FPBinOp ? FPBinOp->getFastMathFlags() : FastMathFlags()));
   }
 
+  VPInstruction *createStepVector(Type *Ty) {
```
Could introducing step vector and updating VpWidenIntOrFpInductionRecipe to not generate it itself be split off separately?

lukel97 replied:
I've opened #129508 for this, thanks

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1757,7 +1760,9 @@ class VPWidenInductionRecipe : public VPHeaderPHIRecipe {
 
 /// A recipe for handling phi nodes of integer and floating-point inductions,
-/// producing their vector values.
+/// producing their vector values. This won't execute any LLVM IR and will get
+/// expanded later into VPWidenIntOrFpInitialRecipe, VPWidenIntOrFpPHIRecipe and
+/// VPWidenIntOrFpBackedgeRecipe.
```
Are the 3 new recipes still in this patch?

lukel97 replied:
Oops no, I've removed the references

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1938,11 +1944,16 @@ class VPScalarPHIRecipe : public VPHeaderPHIRecipe {
-  /// debug location \p DL.
-  VPWidenPHIRecipe(PHINode *Phi, VPValue *Start = nullptr, DebugLoc DL = {})
-      : VPSingleDefRecipe(VPDef::VPWidenPHISC, ArrayRef<VPValue *>(), Phi, DL) {
+  /// debug location \p DL..
```
```suggestion
  /// debug location \p DL.
```

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2029,10 +2029,146 @@ void VPlanTransforms::createInterleaveGroups(
+    // floating-point arithmetic as appropriate.
+    if (Ty->isFloatingPointTy())
+      VF = Builder.createScalarCast(Instruction::CastOps::UIToFP, VF, Ty, DL);
+    else if (Ty != TypeInfo.inferScalarType(VF))
+      VF = Builder.createScalarCast(Instruction::CastOps::Trunc, VF, Ty, DL);
```
If `Ty != TypeInfo.inferScalarType(VF)`, why is it always truncated instead of extended?

lukel97 replied:
I originally also had code to extend it, but it turns out this is never reached so I removed it: https://github.com/llvm/llvm-project/pull/118638/commits/e7f5a4eb653687585c765e9e8488313dfd0d548c

There's currently an invariant that the VF passed in is at least as wide as the type of the IV. Today the VF is always the canonical IV type, but this might change when we replace the VF with the runtime EVL since @get.vector.length always returns i32. I was planning on adding the extend case when we get to that patch, what do you think?

See also here: https://github.com/llvm/llvm-project/pull/119284/files#diff-53267225b83e943ceae51c326c9941e323fd7aaf74a08b5e6998d6456f88d1ddR2097-R2102

lukel97 replied:
This has been removed now to more closely match the original code

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1938,11 +1944,16 @@ class VPScalarPHIRecipe : public VPHeaderPHIRecipe {
-      : VPSingleDefRecipe(VPDef::VPWidenPHISC, ArrayRef<VPValue *>(), Phi, DL) {
+  /// debug location \p DL..
+  VPWidenPHIRecipe(Instruction *Phi, VPValue *Start = nullptr, DebugLoc DL = {},
+                   const Twine &Name = "vec.phi")
+      : VPSingleDefRecipe(VPDef::VPWidenPHISC, ArrayRef<VPValue *>(), Phi, DL),
+        Name(Name.str()) {
```
Agreed, I've split this off into #129527

lukel97 committed:
Update vplan test to include preheader, remove references to old recipes, remove stray .

lukel97 commented on llvm/test/Transforms/LoopVectorize/vplan-iv-transforms.ll:
```diff
@@ -65,3 +65,57 @@ loop.latch:
+; CHECK-NEXT:     vp<%6> = vector-pointer ir<%q>
+; CHECK-NEXT:     WIDEN store vp<%6>, ir<%y>
+; CHECK-NEXT:     EMIT vp<%index.next> = add nuw vp<%3>, ir<8>
+; CHECK-NEXT:     EMIT vp<%vec.ind.next> = add ir<%i>, vp<%2>
```
Yeah, I've added the preheader to the test now (and added some regexes for the value names)

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

lukel97 committed:
Update now that #129508 is landed

* Remove temp stepvector operand from VPWidenIntOrFpInductionRecipe
* Use VPInstruction::Broadcast

lukel97 committed:
clang-format

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -966,6 +966,7 @@ bool VPInstruction::onlyFirstLaneUsed(const VPValue *Op) const {
   case VPInstruction::BranchOnCount:
   case VPInstruction::BranchOnCond:
   case VPInstruction::ResumePhi:
+  case VPInstruction::Broadcast:
```
This was needed since we're now using VPInstruction::Broadcasts for arbitrary VPValues

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1086,15 +1087,14 @@ void VPInstruction::print(raw_ostream &O, const Twine &Indent,
-  switch (getOpcode()) {
-  case Instruction::ZExt:
-  case Instruction::Trunc: {
+  if (isScalarCast()) {
```
Needed because one of the recipes expanded to is Instruction::UIToFP. Happy to split this off to an NFC if wanted.

lukel97 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2762,8 +2762,7 @@ LoopVectorizationCostModel::getVectorIntrinsicCost(CallInst *CI,
   // Fix widened non-induction PHIs by setting up the PHI operands.
-  if (EnableVPlanNativePath)
-    fixNonInductionPHIs(State);
+  fixNonInductionPHIs(State);
```
Needed because we're now using VPWidenPHI on the non-vplan-native path. Happy to split this off into an NFC as well if wanted

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2390,6 +2380,129 @@ void VPlanTransforms::createInterleaveGroups(
+  if (StepTy->isFloatingPointTy())
+    Init = Builder.createWidenCast(Instruction::UIToFP, Init, StepTy);
+
+  VPValue *SplatStart = Builder.createNaryOp(VPInstruction::Broadcast, Start);
+  VPValue *SplatStep = Builder.createNaryOp(VPInstruction::Broadcast, Step);
```
Yeah materializeBroadcasts runs before convertToConcrete recipes, but it also only works on live ins and values defined in the entry block. The increment is calculated in the loop and it needs to be broadcast too

lukel97 committed:
Fix typo, use TypeInfo in assert

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2398,6 +2511,20 @@ void VPlanTransforms::convertToConcreteRecipes(VPlan &Plan,
   for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
            vp_depth_first_deep(Plan.getEntry()))) {
+
+    // Move VPWidenPointerInductionRecipes to the back of the phis
+    // since it may insert non-phi instructions in place, which will
+    // interfere with other header phis if they come after.
+    //
+    // TODO: Expand out VPWidenPointerInductionRecipe into multiple
+    // recipes here and remove this
```
I don't think firstNonPhi is enough, since any VPWidenIntOrFpInductionRecipe that gets executed before a VPWidenPointerInductionRecipe will end up generating something like:

```llvm
  %pointer.phi = phi ptr [ null, %vector.ph ], [ %ptr.ind, %vector.ph ]
  %ptr.ind = getelementptr i8, ptr %pointer.phi, i64 16
  %vector.gep = getelementptr i8, ptr %pointer.phi, <4 x i64> <i64 0, i64 4, i64 8, i64 12>
  %vec.ind = phi <4 x i32>  ; VPWidenIntOrFpInductionRecipe's VPWidenPHI
```

Alternatively, we could tweak VPWidenPHI to always insert its phi at firstNonPhi during execution?

lukel97 replied:
It looks like fixing up the location in VPWidenPHI actually removes some test diffs, so I've gone ahead and added that in ec5fe59bf46e7a81362165b68cb64d0701f2a2ff

lukel97 committed:
Don't move VPWidenPointerInductionRecipe, instead fixup location of phi in VPWidenPHI::execute

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2390,6 +2380,129 @@ void VPlanTransforms::createInterleaveGroups(
+  // If the phi is truncated, truncate the start and step values.
+  VPBuilder Builder(Plan->getVectorPreheader());
+  if (isa<TruncInst>(IV)) {
+    assert(Start->getUnderlyingValue()->getType()->isIntegerTy() &&
```
Done in 61bd641

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

lukel97 committed:
Fix comments

lukel97 commented on llvm/test/Transforms/LoopVectorize/AArch64/transform-narrow-interleave-to-widen-memory-remove-loop-region.ll:
```diff
@@ -109,8 +109,8 @@ define void @load_store_interleave_group_tc_2(ptr noalias %data) {
 ; VF4:       [[PRED_STORE_CONTINUE6]]:
-; VF4-NEXT:    [[VEC_IND_NEXT]] = add <4 x i8> [[VEC_IND]], splat (i8 4)
 ; VF4-NEXT:    [[INDEX_NEXT]] = add nuw i64 [[INDEX]], 4
+; VF4-NEXT:    [[VEC_IND_NEXT]] = add <4 x i8> [[VEC_IND]], splat (i8 4)
```
I gave this a try by inserting it before the canonical IV increment but it ended up creating more test diffs. It looks like there's a lot of cases where the widened IV increment comes after the canonical IV, e.g. in test/Transforms/LoopVectorize/vector-geps.ll

fhahn replied:
Ah OK, thanks for checking!

lukel97 committed:
Remove VPWidenPHIRecipe change

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2026,7 +2018,7 @@ class VPWidenPHIRecipe : public VPSingleDefRecipe, public VPPhiAccessors {
   /// Create a new VPWidenPHIRecipe for \p Phi with start value \p Start and
   /// debug location \p DL.
-  VPWidenPHIRecipe(PHINode *Phi, VPValue *Start = nullptr, DebugLoc DL = {},
+  VPWidenPHIRecipe(Instruction *Phi, VPValue *Start = nullptr, DebugLoc DL = {},
```
A VPWidenIntOrFpInduction's underlying value can be a trunc instruction too, but now that you mention it I think it's ok if we just pass `WidenIVR->getPHINode()` in here. I've removed it in 993ba23953bf7720562fe3814e38449a2dce3818

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

lukel97 commented on llvm/test/Transforms/LoopVectorize/scalable-inductions.ll:
```diff
@@ -23,12 +23,12 @@ define void @add_ind64_unrolled(ptr noalias nocapture %a, ptr noalias nocapture
 ; CHECK:       vector.body:
 ; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, [[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], [[VECTOR_BODY]] ]
 ; CHECK-NEXT:    [[VEC_IND:%.*]] = phi <vscale x 2 x i64> [ [[TMP6]], [[VECTOR_PH]] ], [ [[VEC_IND_NEXT:%.*]], [[VECTOR_BODY]] ]
+; CHECK-NEXT:    [[BROADCAST_SPLATINSERT:%.*]] = insertelement <vscale x 2 x i64> poison, i64 [[TMP8]], i64 0
+; CHECK-NEXT:    [[DOTSPLAT:%.*]] = shufflevector <vscale x 2 x i64> [[BROADCAST_SPLATINSERT]], <vscale x 2 x i64> poison, <vscale x 2 x i32> zeroinitializer
```
Making a note here that if we're removing the hoisting logic in https://github.com/llvm/llvm-project/pull/142594 anyway then these diffs where the broadcast is sunk shouldn't be important. LICM will hoist them out afterwards anyway

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

lukel97 commented:
Ping, all dependent patches should be landed now

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/split-VPWidenIntOrFpInductionRecipe

lukel97 committed:
Infer and check types instead of peeking through underlying value for trunc

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2511,6 +2500,130 @@ void VPlanTransforms::createInterleaveGroups(
+  // The value from the original loop to which we are mapping the new induction
+  // variable.
+  Instruction *IV = Trunc ? cast<Instruction>(Trunc) : WidenIVR->getPHINode();
+  Type *Ty = IV->getType();
```
Done in 7cf85b9091571517d5b97ef84275a3bc377617a8

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2511,6 +2500,130 @@ void VPlanTransforms::createInterleaveGroups(
+
+  // If the phi is truncated, truncate the start and step values.
+  VPBuilder Builder(Plan->getVectorPreheader());
+  if (isa<TruncInst>(IV)) {
```
Good idea, done in 7cf85b9091571517d5b97ef84275a3bc377617a8

fhahn commented on llvm/test/Transforms/LoopVectorize/pointer-induction.ll:
```diff
@@ -282,3 +282,141 @@ for.cond:                                         ; preds = %for.body, %entry
+  br label %for.body
+
+for.body:                                         ; preds = %for.body, %entry
+  %iv.ptr = phi ptr [ null, %entry ], [ %iv.ptr.next, %for.body ]
```
probably best to start at a concrete pointer, to avoid poison.

fhahn reviewed (approved):
LGTM with a few more small comments inline, thanks!

lukel97 committed:
Move comment to flags

lukel97 committed:
Use ToRemove

lukel97 committed:
Remove null ptr in test, update variable names

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2518,6 +2507,128 @@ void VPlanTransforms::createInterleaveGroups(
+
+  WidePHI->addOperand(Next);
+
+  WidenIVR->replaceAllUsesWith(WidePHI);
```
Thanks, done in e8d64a4
